clippy_lints/src/dereference.rs

```diff
@@ -0,0 +1,153 @@
+                                // Caller must be a variable
+                                let variables = arg_list[0];
+                                if variables.len() == 1;
+                                if let ExprKind::Path(QPath::Resolved(None, _)) = variables[0].kind;
+
+                                then {
+                                    let name = method_names[0].as_str();
+                                    lint_deref(cx, &*name, &variables[0], variables[0].span, arg.span);
+                                }
+                            }
+                        }
+                    }
+                    ExprKind::MethodCall(ref method_name, _, ref args) => {
+                        if init.span.from_expansion() {
+                            return;
+                        }
+                        if_chain! {
+                            if args.len() == 1;
+                            if let ExprKind::Path(QPath::Resolved(None, _)) = args[0].kind;
+                            // Caller must call only one other function (deref or deref_mut)
+                            // otherwise it can lead to error prone suggestions (ie: &*a.len())
+                            let (method_names, arg_list, _) = method_calls(init, 2);
+                            if method_names.len() == 1;
+                            // Caller must be a variable
+                            let variables = arg_list[0];
+                            if variables.len() == 1;
+                            if let ExprKind::Path(QPath::Resolved(None, _)) = variables[0].kind;
+
+                            then {
+                                let name = method_name.ident.as_str();
+                                lint_deref(cx, &*name, init, args[0].span, init.span);
+                            }
+                        }
+                    }
+                    _ => ()
+                }
+            }
+        }
+    }
+
+    fn check_expr(&mut self, cx: &LateContext<'a, 'tcx>, expr: &'tcx Expr<'_>) {
+        if_chain! {
+            if let ExprKind::MethodCall(ref method_name, _, ref args) = &expr.kind;
+            if args.len() == 1;
+            if let ExprKind::Path(QPath::Resolved(None, _)) = args[0].kind;
+            if let Some(parent) = get_parent_expr(cx, &expr);
+
+            then {
+                match parent.kind {
+                    // Already linted using statements
+                    ExprKind::Call(_, _) | ExprKind::MethodCall(_, _, _) => (),
+                    _ => {
+                        let name = method_name.ident.as_str();
+                        lint_deref(cx, &*name, &args[0], args[0].span, expr.span);
+                    }
+                }
+            }
+        }
+    }
+}
+
+fn lint_deref(cx: &LateContext<'_, '_>, fn_name: &str, call_expr: &Expr<'_>, var_span: Span, expr_span: Span) {
+    match fn_name {
+        "deref" => {
+            if get_trait_def_id(cx, &paths::DEREF_TRAIT)
```

Comment:
```suggestion
            if cx.tcx.lang_items().deref_trait()
```

same with `deref_mut_trait` below: https://doc.rust-lang.org/nightly/nightly-rustc/rustc/middle/lang_items/struct.LanguageItems.html#method.deref_trait